docs/proposals/scaled-prometheus-pipeline/scaled-prometheus-pipeline.md

```diff
@@ -0,0 +1,139 @@
+
+![newMetricsPipeline.png](./newMetricsPipeline.png)
+
+To avoid making performance worse, we of course need to ensure that for each series, it is always sent to the same pushgateway and prometheus server. To do that, we need to implement some form of hashing. Preferably, this would all be done on the controllers, to avoid having to introduce a new “distributor” component. 
+
+### Distributing Metrics
+
+On prod, out of the ~105k metric series, only about 1.5k do not belong to a gateway. The average number of metric series per gateway is only ~350. This makes `gatewayID` a useful key to hash on. We simply map each `gatewayID` to a specific pushgateway, and all non-gateway metrics also to a single specific pushgateway. Even a very simple implementation with this method will ensure that no server has more than about 1000 series more than average. Importantly, all of this work can be done on the controller in the `Exporter`. 
+
+**Discussion:**
+While hashing on the controller provides an easy solution which is already scalable, it may be more architecturally sound to instead shift this responsibility to a load balancer in front of the pushgateways. Unfortunately, we cannot use a naive load-balancer since the metrics must be directed to the correct gateway. It would be possible to modify the pushgateway as a load balancer with this functionality. Simply configure it to hash the metrics as they come in and send them to the correct pushgateway to be scraped. However we then are left with another bottleneck to deal with. For this reason, I believe hashing on the controllers is the best option.
+
+### Dealing with Rescaling
+
+When a user decides to rescale their deployment (either up or down), series will inevitably be mapped to different pushgateways. This will mean that series will be scraped by different prometheus servers as well. Note that this will not cause any issues when querying/alerting as Thanos can handle the same series existing across multiple servers. The only effect this will have is some temporary non-optimal distribution of series since some will exist in two locations. Once the local server retention time has passed, everything will be back to normal, but even during that period we expect no noticeable impact to performance or user experience.
+
+We will use consistent hashing on the exporters to ensure that as few series as possible are moved.
+
+### Pushgateway Redundancy
+
+We can leverage the addition of multiple gateways to provide failure protection. As discussed above, it's not a significant problem if some series get shifted to a different prometheus server. Because of this, we can protect metrics against a pushgateway failure by simply pushing metrics to the next gateway in the hash circle if the first request fails. While I do not expect this to happen much if at all (as far as I know a pushgateway has never failed while in prod/staging), this provides extra protection at no cost.
```

Comment:
A mitigation for this would be to use a queue service (e.g. Amazon SQS, Apache Kafka) to provide some fault tolerance if pushgateways go down.

Reply:
Could definitely be something worth looking into. Would need to weigh the added complexity and costs of course. At first glance I feel like it probably wouldn't be worth it since the problem it solves (small number of datapoints in a different server) would be minimal at most but I'll keep it in mind.